File: src/js/main/RichJsonModule.js
import {__RICH_JSON_COMMANDS, __RICH_JSON_LATE_APPLIES} from "./RichJsonCommandHolder.js";

const __RICH_JSON_MODULES = {};

/**
 * Use this class in order to write your own RichJson modules.
 */
export class RichJsonModule {
    name = "";
    lateApplies = [];
    commands = {};
    kcmdIgnores = {};
    isIncluded = false;

    constructor(name) {
        this.name = name;
    }

    /**
     * Registers a command that is flagged for late application.
     * This adds the command name to the late application queue and defines its
     * execution logic and ignore rules.
     *
     * @param name    The unique identifier for the command.
     * @param func    The function or logic to be executed when the command is called.
     * @param ignores An optional array of identifiers or contexts that this command
     * should ignore. Defaults to undefined.
     * @return        The current instance for method chaining.
     */
    addLateApply(name, func, ignores = undefined) {
        this.lateApplies.push(name);
        if (Array.isArray(ignores)) {
            this.kcmdIgnores[name] = ignores;
        }
        return this.addCommand(name, func);
    }

    /**
     * Adds a standard command to the command registry.
     * Mapping the command name to its respective function and optionally
     * setting up ignore rules.
     *
     * @param name    The unique identifier for the command.
     * @param func    The function or logic to be executed when the command is called.
     * @param ignores An optional array of identifiers or contexts that this command
     * should ignore. Defaults to undefined.
     * @return        The current instance for method chaining.
     */
    addCommand(name, func, ignores = undefined) {
        this.commands[name] = func;
        if (Array.isArray(ignores)) {
            this.kcmdIgnores[name] = ignores;
        }
        return this;
    }

    __include() {
        this.isIncluded = true;
        Object.entries(this.commands).forEach(([name, func]) => {
            if (Object.hasOwn(__RICH_JSON_COMMANDS.built_in, name)) {
                throw (`RichJson you can not override built in commands. Affected command is '#${name}'`);
            }
            __RICH_JSON_COMMANDS.available[name] = func;
            __RICH_JSON_COMMANDS.enabled[name] = func;
        });
        this.lateApplies.forEach((name) => {
            if (!__RICH_JSON_LATE_APPLIES.includes(name)) {
                __RICH_JSON_LATE_APPLIES.push(name);
            }
        });
        Object.entries(this.kcmdIgnores).forEach(([name, ignored]) => {
            __RICH_JSON_COMMANDS.kcmd_ignored[name] = ignored;
        })
    }

    __exclude() {
        this.isIncluded = false;
        Object.entries(this.lateApplies).forEach(([name]) => {
            __RICH_JSON_LATE_APPLIES.splice(__RICH_JSON_LATE_APPLIES.indexOf(name), 1);
        });
        Object.entries(this.commands).forEach(([name]) => {
            delete __RICH_JSON_COMMANDS.available[name];
            delete __RICH_JSON_COMMANDS.enabled[name];
        });
    }
}

/**
 * Registers a RichJson module.
 * @param module
 * @returns {*}
 */
export function registerModule(module) {
    console.log(`RichJson registering module '${module.name}'`);
    __RICH_JSON_MODULES[module.name] = module;
    return module;
}

/**
 * Unregister a RichJson module.
 * @param name
 */
export function unregisterModule(name) {
    if (isModuleRegistered(name)) {
        if (__RICH_JSON_MODULES[name].isIncluded)
            throw (`RichJson can not unregister module '${name}' due to it is currently included`);
        console.log(`RichJson unregistering module '${name}'`);
        delete __RICH_JSON_MODULES[name];
    }
}

/**
 * Checks if a RichJson module is registered.
 * @param name
 * @returns {boolean}
 */
export function isModuleRegistered(name) {
    return Object.hasOwn(__RICH_JSON_MODULES, name);
}

/**
 * Includes a RichJson module.
 * @param name
 */
export function includeModule(name) {
    if (isModuleRegistered(name)) {
        if (!__RICH_JSON_MODULES[name].isIncluded) {
            console.log(`RichJson including module '${name}'`);
            __RICH_JSON_MODULES[name].__include();
        }
    }
}

/**
 * Excludes a RichJson module.
 * @param name
 */
export function excludeModule(name) {
    if (isModuleRegistered(name)) {
        if (__RICH_JSON_MODULES[name].isIncluded) {
            console.log(`RichJson excluding module '${name}'`);
            __RICH_JSON_MODULES[name].__exclude();
        }
    }
}

File: src/python/main/module/rich_json_module.py
import logging
from ..core.rich_json import _RICH_JSON_COMMANDS
from ..core.rich_json_constants import _RICH_JSON_LATE_APPLIES

# Globales Dictionary für registrierte Module
_RICH_JSON_MODULES = {}

class RichJsonModule:
    """
    Use this class in order to write your own RichJson modules.
    """
    def __init__(self, name):
        self.name = name
        self.late_applies = []
        self.commands = {}
        self.kcmd_ignores = {}
        self.is_included = False

    def add_late_apply(self, name, func, ignores=None):
        """
        Registers a command that is flagged for late application.
        This adds the command name to the late application queue and defines its
        execution logic and ignore rules.

        :param name: The unique identifier for the command.
        :param func: The function or logic to be executed when the command is called.
        :param ignores: An optional list of identifiers or contexts that this command should ignore.
        :return: The current instance for method chaining.
        """
        self.late_applies.append(name)
        if isinstance(ignores, list):
            self.kcmd_ignores[name] = ignores

        return self.add_command(name, func)

    def add_command(self, name, func, ignores=None):
        """
        Adds a standard command to the command registry.
        Mapping the command name to its respective function and optionally
        setting up ignore rules.

        :param name: The unique identifier for the command.
        :param func: The function or logic to be executed when the command is called.
        :param ignores: An optional list of identifiers or contexts that this command should ignore.
        :return: The current instance for method chaining.
        """
        self.commands[name] = func
        if isinstance(ignores, list):
            self.kcmd_ignores[name] = ignores

        return self

    def _include(self):
        """
        Internal method to include the module's commands into the global registry.
        """
        self.is_included = True

        for name, func in self.commands.items():
            if name in _RICH_JSON_COMMANDS.built_in:
                raise ValueError(f"RichJson: You cannot override built-in commands. Affected command is '#{name}'")

            _RICH_JSON_COMMANDS.available[name] = func
            _RICH_JSON_COMMANDS.enabled[name] = func

        for name in self.late_applies:
            if name not in _RICH_JSON_LATE_APPLIES:
                _RICH_JSON_LATE_APPLIES.append(name)

        for name, ignored in self.kcmd_ignores.items():
            _RICH_JSON_COMMANDS.kcmd_ignored[name] = ignored

    def _exclude(self):
        """
        Internal method to remove the module's commands from the global registry.
        """
        self.is_included = False

        for name in self.late_applies:
            if name in _RICH_JSON_LATE_APPLIES:
                _RICH_JSON_LATE_APPLIES.remove(name)

        for name in self.commands.keys():
            _RICH_JSON_COMMANDS.available.pop(name, None)
            _RICH_JSON_COMMANDS.enabled.pop(name, None)


def register_module(module):
    """
    Registers a RichJson module.
    """
    logging.info(f"RichJson registering module '{module.name}'")
    _RICH_JSON_MODULES[module.name] = module
    return module


def unregister_module(name):
    """
    Unregister a RichJson module.
    """
    if is_module_registered(name):
        module = _RICH_JSON_MODULES[name]
        if module.is_included:
            raise RuntimeError(f"RichJson: Cannot unregister module '{name}' because it is currently included")

        logging.info(f"RichJson unregistering module '{name}'")
        del _RICH_JSON_MODULES[name]


def is_module_registered(name):
    """
    Checks if a RichJson module is registered.
    """
    return name in _RICH_JSON_MODULES


def include_module(name):
    """
    Includes a RichJson module.
    """
    if is_module_registered(name):
        module = _RICH_JSON_MODULES[name]
        if not module.is_included:
            logging.info(f"RichJson including module '{name}'")
            module._include()


def exclude_module(name):
    """
    Excludes a RichJson module.
    """
    if is_module_registered(name):
        module = _RICH_JSON_MODULES[name]
        if module.is_included:
            logging.info(f"RichJson excluding module '{name}'")
            module._exclude()